src/pkg/egress/syslog/https_batch.go

```diff
@@ -13,15 +13,120 @@ import (
 	"github.com/valyala/fasthttp"
 )
 
+// --- Coordinator definition ---
+type RetryCoordinator struct {
+	sem chan struct{}
+}
+
+var (
+	globalRetryCoordinator     *RetryCoordinator
+	globalRetryCoordinatorOnce sync.Once
+	maxParallelRetries         = 2
+)
+
+func WithParallelRetries(n int) {
+	maxParallelRetries = n
+	globalRetryCoordinatorOnce = sync.Once{}
+	globalRetryCoordinator = nil
+}
+
+func GetGlobalRetryCoordinator() *RetryCoordinator {
+	globalRetryCoordinatorOnce.Do(func() {
+		globalRetryCoordinator = &RetryCoordinator{
+			sem: make(chan struct{}, maxParallelRetries),
+		}
+	})
+	return globalRetryCoordinator
+}
+
+func (c *RetryCoordinator) Acquire() {
+	c.sem <- struct{}{}
+}
+
+func (c *RetryCoordinator) Release() {
+	<-c.sem
+}
+
+type InternalRetryWriter interface {
+	ConfigureRetry(retryDuration RetryDuration, maxRetries int)
+}
+
+type Retryer struct {
+	retryDuration RetryDuration
+	maxRetries    int
+	binding       *URLBinding
+	coordinator   *RetryCoordinator
+}
+
+func NewRetryer(
+	binding *URLBinding,
+	retryDuration RetryDuration,
+	maxRetries int,
+) *Retryer {
+	return &Retryer{
+		retryDuration: retryDuration,
+		maxRetries:    maxRetries,
+		binding:       binding,
+		coordinator:   GetGlobalRetryCoordinator(),
+	}
+}
+
+func (r *Retryer) Retry(batch []byte, msgCount float64, function func([]byte, float64) error) {
+	logTemplate := "failed to write to %s, retrying in %s, err: %s"
+	var err error
+
+	// First attempt (fast path, not counted as a retry)
+	err = function(batch, msgCount)
+	if err == nil {
+		return
+	}
+
+	if egress.ContextDone(r.binding.Context) {
+		log.Printf("Context cancelled for %s, aborting retries", r.binding.URL.Host)
+		return
+	}
+
+	log.Printf(logTemplate, r.binding.URL.Host, r.retryDuration(0), err)
+
+	// Now acquire a global retry slot for subsequent retries
+	r.coordinator.Acquire()
+	defer r.coordinator.Release()
+
+	for i := 0; i < r.maxRetries-1; i++ {
+		sleepDuration := r.retryDuration(i)
+		time.Sleep(sleepDuration)
```

Comment:
If the context is cancelled while multiple retries are blocked waiting for the `Acquire()` call to return, each retry will still execute `time.Sleep(...)` regardless of the cancelled context. As a result, the total wait time for all `Retry()` method calls to return can accumulate significantly.

My proposal would be to check the context once again before calling `time.Sleep(...)`.

In general, the `time.Sleep()` approach can cause delays, when the context is canceled. Using `select` with `time.After` could be an alternative.